File: Prácticas/P5/JAVA/Civitas/src/civitas/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;

/**
 *
 * @author Jose Padial
 */
public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;
    
    public Tablero(int num){
        
        if(num >=1){
           this.numCasillaCarcel = num;
        }else{
           this.numCasillaCarcel = 1;
        }
        
        this.casillas = new ArrayList<>();
        this.aniadeCasillaSalida();
        
        this.porSalida = 0;
        this.tieneJuez = false;
        
    }
    
    private void aniadeCasillaSalida(){
        casillas.add(new Casilla("Salida"));
    }
    
    private boolean correcto(){
        if(casillas.size() > numCasillaCarcel && tieneJuez == true){
            return true;
        }else{
            return false;
        }
    }
    
    private boolean correcto(int numCasilla){
        if(correcto() == true && numCasilla <= casillas.size()){
            return true;
        }else{
            return false;
        }
    }
    
    int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        return porSalida;
    }
    
    void aniadeCasilla(Casilla casilla){
        if(casillas.size() == numCasillaCarcel){
            casillas.add(new Casilla("Carcel"));
            casillas.add(casilla);
        }else{
            casillas.add(casilla);
            if(casillas.size() == numCasillaCarcel){
                casillas.add(new Casilla("Carcel"));
            }
        }   
    }
    
    void aniadeJuez(){
        if(tieneJuez == false){
            aniadeCasilla(new CasillaJuez("Juez",numCasillaCarcel));
            tieneJuez = true;
        }
    }
    
    Casilla getCasilla(int numCasilla){
        return casillas.get(numCasilla);
    }
    
    int nuevaPosicion(int actual, int tirada){
        int aux = actual+tirada;
        int posNew = 0;
        if(aux >= casillas.size()){
            posNew = aux % casillas.size();
            if(actual > posNew){
            porSalida = porSalida + 1;
            }
            return posNew;
        }
        else{
            return aux;
        }
    }
    
    int calcularTirada(int origen, int destino){
        int aux = destino - origen;
        
        if(aux < 0 ){
            aux = aux + casillas.size();
            return aux;
        }else{
             return aux;       
             }
    }
    
    @Override
    public String toString() {
        return "Tablero{" + "casillas=" + casillas + ", carcel=" + numCasillaCarcel + '}';
    }
}

File: Prácticas/P4/RUBY/Civitas/lib/tablero.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


require_relative"casilla.rb"
require_relative"diario.rb"


module Civitas
  class Tablero
  
    def initialize(num)
      if(num>1)
        @numCasillaCarcel = num
      else
        @numCasillaCarcel = 1
      end
      
      @casillas = Array.new
      @casillas << Casilla.new("Salida")
      @porSalida = 0
      @tieneJuez = false
    end
    
    def aniade_casilla_salida
      @casillas << Casilla.init_descanso("Salida")
    end
    
    def correctoSA
      if(@casillas.length > @numCasillaCarcel && @tieneJuez == true)
        return true
      else
        return false
      end
    end
    
    def correcto(numCasilla)
      if(correctoSA == true && numCasilla <= @casillas.length)
        return true
      else
        return false
      end
    end
    
    def get_carcel
      @numCasillaCarcel
    end

    def get_por_salida
      @porSalida
    end

    def aniade_casilla(casilla)
      if(@casillas.length == @numCasillaCarcel)
        @casillas << Casilla.init_descanso("Carcel")
      end
      
      @casillas << casilla
      
      if(@casillas.length == @numCasillaCarcel)
        @casillas << Casilla.new("Carcel")
      end
      
    end
    
    def aniade_juez
      if(@tieneJuez == false)
        CasillaJuez.new("Juez",@numCasillaCarcel)
        @tieneJuez = true
      end
    end
    
    def get_casilla(num)
      if(num <= @casillas.length)
        return @casillas.at(num)
      else
        return nil
      end
    end
    
    def nueva_posicion(actual,tirada)
      if(correctoSA == false)
        return -1
      else
        
        pos = actual + tirada
        
        if(pos >= @casillas.length)
          @porSalida = @porSalida + 1
          newpos = pos - @casillas.length
          return newpos
        else
          return pos
        end
        
      end
    end
    
    def calcular_tirada(origen,destino)
      tirada = destino - origen
      
      if(tirada <= 0)
        newtirada = tirada + @casillas.length
        return newtirada
      else
        return tirada
      end
    end

    def to_s
       to_s = "Tablero ( casillas = #{@casillas} carcel = #{@numCasillaCarcel}"
    end
    
  end
end